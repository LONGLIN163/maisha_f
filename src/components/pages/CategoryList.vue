<template>
    <div>
        <div class="navbar-div">
            <van-nav-bar data-test="CategoryTitle"  title="Category" /> 
        </div>
        <div>
            <van-row>
                <van-col span="6">
                    <div id="leftNav">
                        <ul>
                            <li 
                                v-for="(item,index) in category" 
                                :key="index" 
                                @click="clickCategory(index,item.ID)" 
                                :class="{categoryActive:activeCategoryIndex==index}"
                            >
                                {{item.MALL_CATEGORY_NAME}}
                            </li>
                        </ul>
                    </div>
                </van-col>
                <van-col span="18">
                    <div class="tabCategorySub">
                        <van-tabs v-model="active" @click="onClickCategorySub">
                            <van-tab 
                               v-for="(item, index) in categorySub" 
                               :key="index" 
                               :title="item.MALL_SUB_NAME" 
                               @click="clickSubCategory(index,item.ID)"
                               :class="{subCategoryActive:activeSubCategoryIndex==index}"
                            >

                            </van-tab>
                        </van-tabs>
                    </div>
                    <!-- product show -->
                    <div id="list-div">
                        <van-pull-refresh v-model="isRefresh" @refresh="onRefresh">

                            <van-list
                                v-model="loading"
                                :finished="finished"
                                @load="onLoad"
                            >
                                <div class="list-item" @click="goGoods(item.ID)" v-for="(item,index) in goodList" :key="index">
                                    <div class="list-item-img"><img :src="item.IMAGE1" width="100%" :onerror="errorImg"/></div>
                                    <div class="list-item-text">
                                        <div>{{item.NAME}}</div>
                                        <div class="">{{item.ORI_PRICE | moneyFilter}}</div>
                                    </div>
                                </div>
                            </van-list>

                        </van-pull-refresh>
                    </div>
                </van-col>
            </van-row>
        </div>
    </div>
</template>

<script>
    import axios from "axios" 
    import serviceApi from '@/serviceApi.config.js'
    import {toMoney} from '@/filters/moneyFilter.js'
    import {Toast} from 'vant'

    export default {
        data() {
            return {
                category:[],
                activeCategoryIndex:0, 
                activeSubCategoryIndex:0,
                active:0,
                categorySub:[],
                loading:false,   //vant loading status
                finished:false,  //vant loading done or not
                isRefresh:false,
                page:1, // init page number 1 
                goodList:[],   
                categorySubId:'', 
                errorImg:'this.src="' + require('@/assets/images/errorImg.png') + '"',  
                categoryId:1

            }       
        },
        created(){
            this.getCategory();
        },
        methods:{
            getGoodList(){
                    axios({
                    url:serviceApi.getGoodsListByCategorySubID,
                    method:'post',
                    data:{
                            categorySubId:this.categorySubId,
                            page:this.page
                        }
                })
                .then(res=>{
                    console.log(res)
                    if(res.data.code == 200 && res.data.message.length ){
                        this.page++
                        this.goodList=this.goodList.concat(res.data.message)
                    }else{
                            this.finished = true;  
                    }
                    this.loading=false;
                    console.log(this.finished)
                })
                .catch(error=>{
                    console.log(error)
                })
            },
            onClickCategorySub(index){
                this.categorySubId= this.categorySub[index].ID
                this.goodList=[]
                this.activeSubCategoryIndex=index
                this.finished = false
                this.page=1
                this.onLoad()
            },
            getCategory() {  
                axios({
                    url:serviceApi.getCategoryList,
                    method:'get',
                })
                .then(res=>{
                    console.log("res---init---",res)
                    if(res.data.code == 200 && res.data.message ){
                        this.category=res.data.message.sort(function(a, b) { 
                            return - ( b.ID - a.ID );
                        });
                        console.log(" this.category---------",this.category)

                        this.getSubCategoryByCategoryId(this.category[0].ID) // get the first sub cat when we get big category
                    }else{
                        Toast('Server error, fetch data failed!')
                    }
                })
                .catch(error=>{
                    console.log(error)
                }) 
            },
            clickCategory(index,categoryId){
                console.log(index)
                this.activeCategoryIndex=index

                // init paging info
                this.page=1
                this.finished = false
                this.goodList=[]

                this.getSubCategoryByCategoryId(categoryId)

            },
            getSubCategoryByCategoryId(categoryId){
                axios({
                    url:serviceApi.getSubCategoryList,
                    method:'post',
                    data:{categoryId:categoryId}
                })
                .then(res=>{
                    console.log("res---sub---",res)
                    if(res.data.code == 200 && res.data.message ){
                        this.categorySub=res.data.message
                        this.active = 0 // active first subcat when switch big category

                        console.log("res---categorySub---",this.categorySub)

                        this.categorySubId=this.categorySub[0].ID
                        this.onLoad()
                    }else{
                        Toast('Server error, fetching data failed!')
                    }  
                })
                .catch(error=>{
                    console.log(error)
                }) 
            },
            onLoad(){
                setTimeout(()=>{
                    this.categorySubId=this.categorySubId ? this.categorySubId : this.categorySub[0].ID
                    this.getGoodList()
                },1000)
            },
            onRefresh(){
                setTimeout(() => {
                    this.isRefresh = false;
                    this.finished = false;
                    this.goodList=[];
                    this.page=1;
                    this.onLoad();
                }, 500);
            },
            goGoods(id){
                this.$router.push({name:'Goods',params:{goodsId:id}})
            }
        },
        filters:{
            moneyFilter(money){
                return toMoney(money)
            }  
        },
    }
</script>

<style lang="scss" scoped>
    #leftNav ul li {
        line-height: 2rem;
        border-bottom:$color-divider2;
        padding:3px;
        font-size:0.8rem;
        text-align: center;
        .categoryActive{
            background-color: $color-background-sub;
        }
    }

    // ***this is vant bug***
    .van-tab--active{
        color:rgb(45, 57, 235) !important;
    }

        
    #list-div{
        .list-item{
            display: flex;
            flex-direction: row;
            font-size:0.8rem;
            border-bottom: 1px solid $color-background;
            background-color: $color-background-sub;
            padding:5px;
            .list-item-img{
                flex:8;
            }
            .list-item-text{
                flex:16;
                margin-top:10px;
                margin-left:10px;
            }
        }
        overflow: scroll;
    }


</style>