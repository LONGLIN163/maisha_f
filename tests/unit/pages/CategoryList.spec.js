import { config, mount, createLocalVue} from '@vue/test-utils'
import CategoryList from '@/components/pages/CategoryList.vue'

// use old syntax for mount method
config.showDeprecationWarnings = false

const Vant=require("vant")

const localVue = createLocalVue();
localVue.use(Vant); 


describe('CategoryList',  () => {

  it('renders CategoryList',async () => {

    const wrapper = mount(CategoryList, {
      localVue
    })

    expect(wrapper).toMatchSnapshot()

  })
})