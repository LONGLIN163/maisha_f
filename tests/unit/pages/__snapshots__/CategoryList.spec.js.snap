// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList renders CategoryList 1`] = `
<div>
  <div class="navbar-div">
    <div class="van-nav-bar van-hairline--bottom">
      <div class="van-nav-bar__content">
        <div class="van-nav-bar__title van-ellipsis">Category</div>
      </div>
    </div>
  </div>
  <div>
    <div class="van-row">
      <div class="van-col van-col--6">
        <div id="leftNav">
          <ul></ul>
        </div>
      </div>
      <div class="van-col van-col--18">
        <div class="tabCategorySub">
          <div class="van-tabs van-tabs--line">
            <div class="van-tabs__wrap">
              <div role="tablist" class="van-tabs__nav van-tabs__nav--line">
                <div class="van-tabs__line"></div>
              </div>
            </div>
            <div class="van-tabs__content"></div>
          </div>
        </div>
        <div id="list-div">
          <div class="van-pull-refresh">
            <div class="van-pull-refresh__track" style="transition-duration: 0ms;">
              <div class="van-pull-refresh__head"></div>
              <div role="feed" class="van-list">
                <div class="van-list__placeholder"></div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
